AlphaWallet/KeyManagement/EtherKeystore.swift

```diff
@@ -0,0 +1,647 @@
+        if let result = (try? LegacyFileBasedKeystore())?.export(privateKey: key, newPassword: newPassword) {
+            completion(result)
+        } else {
+            completion(.failure(.failedToExportPrivateKey))
+        }
+    }
+
+    func exportSeedPhraseOfHdWallet(forAccount account: EthereumAccount, reason: KeystoreExportReason, completion: @escaping (Result<String, KeystoreError>) -> Void) {
+        if let seedPhrase = getSeedPhraseForHdWallet(forAccount: account, prompt: reason.prompt, withUserPresence: isUserPresenceCheckPossible) {
+            completion(.success(seedPhrase))
+        } else {
+            completion(.failure(.failedToExportSeed))
+        }
+    }
+
+    func verifySeedPhraseOfHdWallet(_ inputSeedPhrase: String, forAccount account: EthereumAccount, completion: @escaping (Result<Bool, KeystoreError>) -> Void) {
+        if let actualSeedPhrase = getSeedPhraseForHdWallet(forAccount: account, prompt: R.string.localizable.keystoreAccessKeyHdVerify(), withUserPresence: isUserPresenceCheckPossible) {
+            let matched = inputSeedPhrase.lowercased() == actualSeedPhrase.lowercased()
+            completion(.success(matched))
+        } else {
+            completion(.failure(.failedToExportSeed))
+        }
+    }
+
+    @discardableResult func delete(wallet: Wallet) -> Result<Void, KeystoreError> {
+        switch wallet.type {
+        case .real(let account):
+            //TODO not the best way to do this but let's see if there's a better way to inform the coordinator that a wallet has been deleted
+            PromptBackupCoordinator(keystore: self, wallet: wallet, config: .init()).deleteWallet()
+
+            removeAccountFromBookkeeping(account)
+            deleteKeysAndSeedCipherTextFromKeychain(forAccount: account)
+            deletePrivateKeysFromSecureEnclave(forAccount: account)
+        case .watch(let address):
+            removeAccountFromBookkeeping(.init(address: address))
+        }
+        (try? LegacyFileBasedKeystore())?.delete(wallet: wallet)
+        return .success(())
+    }
+
+    private func deletePrivateKeysFromSecureEnclave(forAccount account: EthereumAccount) {
+        let secureEnclave = SecureEnclave()
+        secureEnclave.deletePrivateKeys(withName: encryptionKeyForPrivateKeyLabel(fromAccount: account, withUserPresence: true))
+        secureEnclave.deletePrivateKeys(withName: encryptionKeyForPrivateKeyLabel(fromAccount: account, withUserPresence: false))
+        secureEnclave.deletePrivateKeys(withName: encryptionKeyForSeedLabel(fromAccount: account, withUserPresence: true))
+        secureEnclave.deletePrivateKeys(withName: encryptionKeyForSeedLabel(fromAccount: account, withUserPresence: false))
+    }
+
+    private func deleteKeysAndSeedCipherTextFromKeychain(forAccount account: EthereumAccount) {
+        keychain.delete("\(Keys.ethereumRawPrivateKeyUserPresenceNotRequiredPrefix)\(account.address.eip55String)")
```

Comment:
Can exceptions be thrown here? This would prevent deletion in the subsequent lines. Make all deletions not contingent on each other, i.e., delete as much as possible rather than "aborting" if a preceding operation fails.

Reply:
No. As above.